src/app/clusters/access-control-server/access-control-server.cpp

```diff
@@ -565,72 +193,54 @@ CHIP_ERROR AccessControlAttribute::WriteAcl(const ConcreteDataAttributePath & aP
 {
     FabricIndex accessingFabricIndex = aDecoder.AccessingFabricIndex();
 
+    size_t oldCount;
+    ReturnErrorOnFailure(GetAccessControl().GetEntryCount(accessingFabricIndex, oldCount));
+    size_t maxCount;
+    ReturnErrorOnFailure(GetAccessControl().GetMaxEntriesPerFabric(maxCount));
+
     if (!aPath.IsListItemOperation())
     {
-        DataModel::DecodableList<AccessControlEntryCodec> list;
+        DataModel::DecodableList<AclStorage::DecodableEntry> list;
         ReturnErrorOnFailure(aDecoder.Decode(list));
 
-        size_t oldCount = 0;
-        size_t allCount;
         size_t newCount;
-        size_t maxCount;
-
-        {
-            AccessControl::EntryIterator it;
-            AccessControl::Entry entry;
-            ReturnErrorOnFailure(GetAccessControl().Entries(it, &accessingFabricIndex));
-            while (it.Next(entry) == CHIP_NO_ERROR)
-            {
-                oldCount++;
-            }
-        }
-
-        ReturnErrorOnFailure(GetAccessControl().GetEntryCount(allCount));
         ReturnErrorOnFailure(list.ComputeSize(&newCount));
-        ReturnErrorOnFailure(GetAccessControl().GetMaxEntryCount(maxCount));
-        VerifyOrReturnError(allCount >= oldCount, CHIP_ERROR_INTERNAL);
-        VerifyOrReturnError(static_cast<size_t>(allCount - oldCount + newCount) <= maxCount,
-                            CHIP_IM_GLOBAL_STATUS(ConstraintError));
+
+        VerifyOrReturnError(newCount <= maxCount, CHIP_IM_GLOBAL_STATUS(ConstraintError));
```

Comment:
Shouldn't this be ResourceExhausted per spec?  "9.11.5.3. ACL Attribute" says:

> An attempt to add an Access Control Entry when no more entries are available SHALL result in a RESOURCE_EXHAUSTED error being reported and the ACL attribute SHALL NOT have the entry added to it. See [access control limits](file:///Users/bzbarsky/connectedhomeip-spec/build/html/index.html#ref_access_control_resource_limits).

```suggestion
        VerifyOrReturnError(newCount <= maxCount, CHIP_IM_GLOBAL_STATUS(ResourceExhausted));
```

Reply:
Fixed in latest revision. I do have to change/fix a few of those error codes in the spec, but this one will remain RESOURCE_EXHAUSTED.